### PR TITLE
Initial implementation

### PR DESCRIPTION
**Pull request by umairidris**

### CODE REVIEW

**Comment by xingao267** [comment]
Maybe worth add a README or example usage for how to use this module at the top level as interfaces, nats and bgp are just list of objects or object.

**Comment by xingao267** [comment]
And some commands how to verify the resources have been created like https://github.com/terraform-google-modules/terraform-google-bastion-host/tree/master/modules/iap-tunneling or https://github.com/terraform-google-modules/terraform-google-bastion-host/tree/master/modules/bastion-group

File: nat.tf [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+  router                             = google_compute_router.router.name
+  region                             = google_compute_router.router.region
+  nat_ip_allocate_option             = lookup(each.value, "nat_ip_allocate_option", length(lookup(each.value, "nat_ips", [])) > 0 ? "MANUAL_ONLY" : "AUTO_ONLY")
+  source_subnetwork_ip_ranges_to_nat = lookup(each.value, "source_subnetwork_ip_ranges_to_nat", "ALL_SUBNETWORKS_ALL_IP_RANGES")
+
+  nat_ips                          = lookup(each.value, "nat_ips", [])
+  min_ports_per_vm                 = lookup(each.value, "min_ports_per_vm", 0)
+  udp_idle_timeout_sec             = lookup(each.value, "udp_idle_timeout_sec", null)
+  icmp_idle_timeout_sec            = lookup(each.value, "icmp_idle_timeout_sec", null)
+  tcp_established_idle_timeout_sec = lookup(each.value, "tcp_established_idle_timeout_sec", null)
```
Comment (xingao267) [inline nat.tf]: The doc is variables.tf says it's default to 1200s. Is it default to 1200s at the resource level so it's set to null here?

Same for other variables here that are default to 30s.

Comment (umairidris) [reply nat.tf]: Added defaults

**Comment by umairidris** [comment]
> And some commands how to verify the resources have been created like https://github.com/terraform-google-modules/terraform-google-bastion-host/tree/master/modules/iap-tunneling or https://github.com/terraform-google-modules/terraform-google-bastion-host/tree/master/modules/bastion-group

There are a lot of resources here so I think it will get out of hand to list all the getters. There also don't seem to be list funcs for some resources like interfaces.

**Comment by umairidris** [comment]
> Maybe worth adding a README or example usage for how to use this module at the top level as interfaces, nats and bgp are just list of objects or object.

Added a simple readme example and a few examples in the examples dir

File: modules/interface/main.tf [hunk 2]
```diff
@@ -0,0 +1,35 @@
+resource "google_compute_router_interface" "interface" {
```
Comment (xingao267) [inline modules/interface/main.tf]: License header in all files.

File: nat.tf (hunk 1 above)
Comment (xingao267) [reply nat.tf]: might be some git issue. the defaults seem not reflected in the pr.

Comment (umairidris) [reply nat.tf]: No, it was my mistake... I only changed it in csr :(

File: modules/interface/main.tf (hunk 2 above)
Comment (umairidris) [reply modules/interface/main.tf]: Done.

File: README.md [hunk 3]
```diff
@@ -55,7 +53,7 @@ The following dependencies must be available:
 A service account with the following roles must be used to provision
 the resources of this module:
 
-- Storage Admin: `roles/storage.admin`
+- Storage Admin: `roles/compute.networkAdmin`
```
Comment (xingao267) [inline README.md]: Network Admin: ...

Comment (umairidris) [reply README.md]: Ops, got tunnel vision there :)

**Comment by umairidris** [comment]
friendly ping

File: README.md [hunk 4]
```diff
@@ -1,10 +1,6 @@
 # terraform-google-cloud-router
 
-This module was generated from [terraform-google-module-template](https://github.com/terraform-google-modules/terraform-google-module-template/), which by default generates a module that simply creates a GCS bucket. As the module develops, this README should be updated.
```
Comment (xingao267) [inline README.md]: Run the generate docs command and update the input/output of this module.

**Pull request merged**